### DISCUSSION TITLE
Bluetooth, assign custom name to Service/Characteristic

### DISCUSSION THREAD

**Discussion by GiacCecc** [micropython/micropython#14129] [Libraries & Drivers]
I‘m working on a project with an ESP32-WROOM-32 that needs the use of Bluetooth to send and receive data from a central (I-Phone) as a GATT-Server. I use the nRF Connect app To interact with the ESP32 and the bluetooth low-level library from Micropython 1.22.2.

The Server has t least two different services with at least one characteristic each.

For the first service and its characteristic I used the "assigned numbers" for Environmental Sensing and Temperature from the SIG (Environmental Sensing Service = 0x181A and Temperature Characteristic = 0x2A6E) and I'm able to receive the data and to see the name of the service and of the characteristic in the nRF Connect app (see Screenshot).

For the second service (STEP_SERV) and its characteristic (STEP_CHAR) I used custom UUIDs. This service should offer the possibility to write an int to the server, that will use this value to rotate a stepper motor.

So far everything works as expected, but I´m having hard times trying to understand how I can assign a custom "name" to the custom service and its characteristic. I would like to see these names in the nRF Connect app when I connect to the server.

I tried adding a "Characteristic User Description" with 0x2901 (STEP_DESC) to the STEP_CHAR, returning an handle (self._desc_step) for it after registration and using this handle to write its name ("STEPPER") with .gatts_write(). Unfortunately it doesn't work.

Here is the code I used:

```
# Stepper Motor Service
STEP_SERV_UUID = bluetooth.UUID("00001001-0000-1000-8000-00805F9B34FB")
STEP_CHAR_UUID = bluetooth.UUID("00001002-0000-1000-8000-00805F9B34FB")

# Characteristic User Description
STEP_DESC = (bluetooth.UUID(0x2901), FLAG_WRITE )

# Characteristic for Stepper
STEP_CHAR = (
    STEP_CHAR_UUID,
    FLAG_READ | FLAG_WRITE | FLAG_NOTIFY,
    (STEP_DESC,)
    )

# Service for Stepper
STEP_SERV = ( STEP_SERV_UUID, (STEP_CHAR,))

# --------------------------------------------------------------------------------------
# Register function
def _register(self):
        
        services = ( ENV_SENSE_SERVICE, STEP_SERV, )
        ( (self._handle_temp,), (self._handle_step, self._desc_step) ) = self._ble.gatts_register_services(services)
        
        NAME = "STEPPER"
        self._ble.gatts_write(self._desc_step, NAME.encode("UTF-8"))
```

I would be glad if someone with some experience with it could give me an hint. Thank you all in advance!!!

![BLE-nRF Connect Screenshot](https://github.com/micropython/micropython/assets/79569028/71f9f323-302c-4597-9926-7719d88b2199)

**Comment by trentp-igor** [comment]
Your descriptor only has write permission.  So the client can set the name, but not read the value.

I also found that the iphone version of nRF Connect will not show the Characteristic User Description (0x2901).  You can see the descriptor with the Android version.

iOS is very restrictive of what apps can do and filters certain characteristics.  For instance, characteristics under 0x1801 Generic Access service can not be read.

**Reply by GiacCecc** [reply]
Thank you for the replay Igor. 
I'll try on Android and maybe come back soon here.
